packages/web-components/src/drawer/drawer.ts

```diff
@@ -0,0 +1,222 @@
+
+  /**
+   * @public
+   * @defaultValue medium
+   * Sets the size of the drawer (small/medium/large).
+   */
+  @attr({ attribute: 'size' })
+  public size: DrawerSize = DrawerSize.medium;
+
+  /**
+   * @public
+   * The dialog element.
+   */
+  @observable
+  public dialog!: HTMLDialogElement;
+
+  /**
+   * @public
+   * Method to emit an event after the dialog's open state changes
+   * HTML spec proposal: https://github.com/whatwg/html/issues/9733
+   */
+  public emitToggle = (): void => {
+    this.$emit('toggle', {
+      oldState: this.dialog.open ? 'closed' : 'open',
+      newState: this.dialog.open ? 'open' : 'closed',
+    });
+  };
+
+  /**
+   * @public
+   * Method to emit an event after the dialog's open state changes
+   * HTML spec proposal: https://github.com/whatwg/html/issues/9733
+   */
+  public emitBeforeToggle = (): void => {
+    this.$emit('beforetoggle', {
+      oldState: this.dialog.open ? 'open' : 'closed',
+      newState: this.dialog.open ? 'closed' : 'open',
+    });
+  };
+
+  /**
+   * @public
+   * Method called when the 'inline' attribute changes
+   */
+  public inlineChanged(oldValue: boolean, newValue: boolean): void {
+    if (this.$fastController.isConnected) {
+      if (newValue) {
+        this.validateConfiguration();
+      }
+    }
+  }
+
+  /**
+   * @public
+   * Method called when the 'type' attribute changes
+   */
+  public typeChanged(oldValue: boolean, newValue: boolean): void {
+    if (this.$fastController.isConnected) {
+      if (newValue != oldValue) {
+        this.validateConfiguration();
+      }
+    }
+  }
+
+  /**
+   * @public
+   * Method to show the drawer
+   */
+  public show(): void {
+    Updates.enqueue(() => {
+      this.emitBeforeToggle();
+      if (this.inline) {
+        this.dialog.show();
+      } else {
+        this.dialog.showModal();
+      }
+      this.emitToggle();
+    });
+  }
+
+  /**
+   * @public
+   * Method to hide the drawer
+   */
+  public hide(): void {
+    this.emitBeforeToggle();
+    this.dialog.close();
+    this.emitToggle();
+  }
+
+  /**
+   * @public
+   * @param event - The click event
+   * @returns boolean - Always returns true
+   * Handles click events on the drawer.
+   */
+  public clickHandler(event: Event): boolean {
+    event.preventDefault();
+    if (this.dialog.open && this.type !== DrawerType.alert && event.target === this.dialog) {
+      this.hide();
+    }
+    return true;
+  }
+
+  /**
+   * @public
+   * @param e - The keydown event
+   * @returns boolean | void
+   * Handles keydown events on the drawer
+   */
+  public keydownHandler = (event: KeyboardEvent): boolean | void => {
+    if (event.defaultPrevented) {
+      return;
+    }
+    switch (event.key) {
+      case keyEscape:
+        event.preventDefault();
+
+        if (this.type !== DrawerType.alert) {
+          this.hide();
+        }
+        break;
+      default:
+        return true;
+    }
+  };
```

Comment:
As a workaround for Safari, I had to remove this in `Dialog` and rely on the `@cancel` event in the template. It was new to me but does handles all the `event.key` logic.